packages/headless-react/src/ssr/client-utils.ts

```diff
@@ -1,23 +1,49 @@
 'use client';
 
-import {useEffect, useReducer, useRef} from 'react';
+import {DependencyList, useEffect, useReducer, useRef} from 'react';
+
+/**
+ * Subscriber is a function that takes a single argument, which is another function `listener` that returns `void`. The Subscriber function itself returns another function that can be used to unsubscribe the `listener`.
+ */
+export type Subscriber = (listener: () => void) => () => void;
+
+export type SnapshotGetter<T> = () => T;
+
+/**
+ * Determine if the given list of dependencies has changed.
+ */
+function useHasDepsChanged(deps: DependencyList) {
+  const ref = useRef<DependencyList | null>(null);
+  if (ref.current === null) {
+    ref.current = deps;
+    return false;
+  }
+  if (
+    ref.current.length === deps.length &&
+    !deps.some((dep, i) => !Object.is(ref.current![i], dep))
+  ) {
+    return false;
+  }
+  ref.current = deps;
+  return true;
+}
 
 /**
  * Alternate for `useSyncExternalStore` which runs into infinite loops when hooks are used in `getSnapshot`
  * https://github.com/facebook/react/issues/24529
  */
 export function useSyncMemoizedStore<T>(
-  subscribe: (listener: () => void) => () => void,
-  getSnapshot: () => T
+  subscribe: Subscriber,
+  getSnapshot: SnapshotGetter<T>
 ) {
-  const stateRef = useRef(getSnapshot());
+  const snapshot = useRef(getSnapshot());
```

Comment:
to avoid `getSnapshot` getting called on every re-render, you should set the initial value here to `null`. Your condition at line 59 (`snapshot.current === null`) should be sufficient to call `getSnapshot` the first time.

Reply:
Great, thanks! Updated. Couldn't use `null` as it was raising type errors when getStaticState is passed in from buildControllerHook. Used `{}` to match getStaticState.